### ISSUE TITLE
4.2.1.rc4: `route` generator action gobbles up trailing newline, breaks 3rd-party generators

### ISSUE THREAD

**Issue by mattbrictson** [rails/rails#19316]
**tl;dr –** In 4.2.1.rc4 `route "resources :people"` no longer appends a newline after the inserted route, which breaks the routes file. This breaks third-party generators like devise. It is a regression from 4.2.0 due to the fix for #15913.

--

Assuming a `routes.rb` that starts out like this:

```ruby
Rails.application.routes.draw do
  root 'welcome#index'
end
```

In Rails 4.2.0, `rails g scaffold foo` produces:

```ruby
Rails.application.routes.draw do
  resources :foos

  root 'welcome#index'
end
```

In Rails 4.2.1 (note the missing newline):

```ruby
Rails.application.routes.draw do
  resources :foos
  root 'welcome#index'
end
```

In terms of the scaffold generator, this is not a problem (the code still works, it is just whitespace that has changed). The reason the code works is because the scaffold generator includes an extra trailing newline when it calls the `route` generator action.

However, the [documentation](http://guides.rubyonrails.org/generators.html#route) for the `route` generator action does not specify that a trailing newline is necessary. The example given in the docs is:

```ruby
route "resources :people"
```

However in 4.2.1 this would produce:

```ruby
Rails.application.routes.draw do
  resources :people  root 'welcome#index'
end
```

No newline separates the new route from existing routes; this causes routing code to break.

This regression is due to the fix for #15913.

Due to the way Thor's `inject_into_file` works, I haven't been able to figure out a fix this regression without reintroducing #15913. My suggestion would be to revert the fix for #15913, since the regression will break third-party generators (I've verified that devise is broken).

**Comment by rafaelfranca** [comment]
cc @pixeltrix

**Comment by sevenseacat** [comment]
This also affects things like Devise, as I just found out when installing Devise into a 4.2.1.rc4 app.

**Comment by arthurnn** [comment]
@sevenseacat @mattbrictson I attempted to fix this on https://github.com/rails/rails/pull/19387 , can you guys give it a try and let me know if it works.
thanks
